scipy/sparse/_dia.py

```diff
@@ -176,30 +176,38 @@ def sum(self, axis=None, dtype=None, out=None):
     sum.__doc__ = _spbase.sum.__doc__
 
     def _add_sparse(self, other):
+        # If other is not DIA format, let them handle us instead.
+        if not isinstance(other, _dia_base):
+            return other._add_sparse(self)
+
+        # Fast path for exact equality of the sparsity structure.
+        if np.array_equal(self.offsets, other.offsets):
+            return self._with_data(self.data + other.data)
+
+        # Find the union of the offsets (which will be sorted and unique).
+        new_offsets = np.union1d(self.offsets, other.offsets)
+        self_idx = np.searchsorted(new_offsets, self.offsets)
+        other_idx = np.searchsorted(new_offsets, other.offsets)
+
+        self_d = self.data.shape[1]
+        other_d = other.data.shape[1]
```

Comment:
Does the shape of `self` and `other` get checked to be equal before this function is called? If they aren't the same shape they would need to broadcast to the same shape. So maybe this could be condensed. ??

Reply:
DIA format is weird, because you can have `a.shape == b.shape` but also `a.data.shape != b.data.shape` (because the length of diagonals you need to store depends on the offsets you're using).